From 65b3f145b0383b55444dc3a95cd4d18ea5655fff Mon Sep 17 00:00:00 2001
From: Sergey Dyasli <sergey.dyasli@citrix.com>
Date: Tue, 26 Mar 2019 09:53:25 +0000
Subject: x86: add XEN_SYSCTL_spec_ctrl and xen-spec-ctrl utility

stop_machine infrastructure is copied from live ucode update.
CPUID level 0x00000007:0.edx is updated on each cpu.
CPUID policies are recalculated if there are any CPUID changes.

Signed-off-by: Sergey Dyasli <sergey.dyasli@citrix.com>

diff --git a/tools/misc/.gitignore b/tools/misc/.gitignore
index 1202c0a72384..2403a2129de1 100644
--- a/tools/misc/.gitignore
+++ b/tools/misc/.gitignore
@@ -1,2 +1,3 @@
 xen-mceinj
+xen-spec-ctrl
 xen-ucode
diff --git a/tools/misc/Makefile b/tools/misc/Makefile
index b025f1839278..d1baf0e076ae 100644
--- a/tools/misc/Makefile
+++ b/tools/misc/Makefile
@@ -24,6 +24,7 @@ INSTALL_SBIN-$(CONFIG_X86)     += xen-lowmemd
 INSTALL_SBIN-$(CONFIG_X86)     += xen-mceinj
 INSTALL_SBIN-$(CONFIG_X86)     += xen-mfndump
 INSTALL_SBIN-$(CONFIG_X86)     += xen-ucode
+INSTALL_SBIN-$(CONFIG_X86)     += xen-spec-ctrl
 INSTALL_SBIN                   += xencov
 INSTALL_SBIN                   += xenlockprof
 INSTALL_SBIN                   += xenperf
@@ -121,4 +122,7 @@ xencov: xencov.o
 xen-ucode: xen-ucode.o
 	$(CC) $(LDFLAGS) -o $@ $< $(LDLIBS_libxenctrl) $(APPEND_LDFLAGS)
 
+xen-spec-ctrl: xen-spec-ctrl.o
+	$(CC) $(LDFLAGS) -o $@ $< $(LDLIBS_libxenctrl) $(APPEND_LDFLAGS)
+
 -include $(DEPS_INCLUDE)
diff --git a/tools/misc/xen-spec-ctrl.c b/tools/misc/xen-spec-ctrl.c
new file mode 100644
index 000000000000..9b958703c98a
--- /dev/null
+++ b/tools/misc/xen-spec-ctrl.c
@@ -0,0 +1,64 @@
+#define _GNU_SOURCE
+
+#include <stdio.h>
+#include <err.h>
+#include <errno.h>
+#include <string.h>
+
+#include <xenctrl.h>
+
+void show_help(void)
+{
+    fprintf(stderr,
+            "xen-spec-ctrl: Xen speculation control tool\n"
+            "Usage: xen-spec-ctrl update\n");
+}
+
+int main(int argc, char *argv[])
+{
+    int ret;
+    struct xen_sysctl sysctl = {0};
+    xc_interface *xch;
+    char *str = "Mitigations have been updated! "
+                "Check xen-cpuid output for the details.";
+
+    if ( argc < 2 || strcmp(argv[1], "update") != 0 )
+    {
+        show_help();
+        return 1;
+    }
+
+    xch = xc_interface_open(NULL, NULL, 0);
+    if ( xch == NULL )
+        err(1, "xc_interface_open");
+
+    sysctl.interface_version = XEN_SYSCTL_INTERFACE_VERSION;
+    sysctl.cmd = XEN_SYSCTL_spec_ctrl;
+    sysctl.u.spec_ctrl.op = XENPF_spec_ctrl_update;
+
+    ret = xc_sysctl(xch, &sysctl);
+    if ( ret != 0 )
+    {
+        switch ( errno )
+        {
+        case ENOEXEC:
+            str = "No new H/W features have been found. "
+                  "Did you forget to update the microcode with xen-ucode?";
+            break;
+
+        case EFAULT:
+            str = "Some error has occured!";
+            break;
+
+        default:
+            str = strerror(errno);
+            break;
+        }
+    }
+
+    fprintf(stderr, "Status: %s\n", str);
+
+    xc_interface_close(xch);
+
+    return 0;
+}
diff --git a/xen/arch/x86/cpuid.c b/xen/arch/x86/cpuid.c
index 76c916647a80..8c1ad237e68e 100644
--- a/xen/arch/x86/cpuid.c
+++ b/xen/arch/x86/cpuid.c
@@ -303,7 +303,7 @@ void cpuid_calculate_raw_policy(void)
     ASSERT(p->x86_vendor == boot_cpu_data.x86_vendor);
 }
 
-static void __init calculate_host_policy(void)
+static void calculate_host_policy(void)
 {
     struct cpuid_policy *p = &host_cpuid_policy;
 
@@ -338,7 +338,7 @@ static void __init calculate_host_policy(void)
     }
 }
 
-static void __init guest_common_feature_adjustments(uint32_t *fs)
+static void guest_common_feature_adjustments(uint32_t *fs)
 {
     /* Unconditionally claim to be able to set the hypervisor bit. */
     __set_bit(X86_FEATURE_HYPERVISOR, fs);
@@ -370,7 +370,7 @@ static void __init guest_common_feature_adjustments(uint32_t *fs)
         __clear_bit(X86_FEATURE_RTM, fs);
 }
 
-static void __init calculate_pv_max_policy(void)
+static void calculate_pv_max_policy(void)
 {
     struct cpuid_policy *p = &pv_max_cpuid_policy;
     uint32_t pv_featureset[FSCAPINTS];
@@ -398,7 +398,7 @@ static void __init calculate_pv_max_policy(void)
     p->extd.raw[0xa] = EMPTY_LEAF; /* No SVM for PV guests. */
 }
 
-static void __init calculate_hvm_max_policy(void)
+static void calculate_hvm_max_policy(void)
 {
     struct cpuid_policy *p = &hvm_max_cpuid_policy;
     uint32_t hvm_featureset[FSCAPINTS];
@@ -459,7 +459,7 @@ static void __init calculate_hvm_max_policy(void)
     recalculate_xstate(p);
 }
 
-void __init init_guest_cpuid(void)
+void init_guest_cpuid(void)
 {
     cpuid_calculate_raw_policy();
     calculate_host_policy();
diff --git a/xen/arch/x86/spec_ctrl.c b/xen/arch/x86/spec_ctrl.c
index 6639a69557a2..0ea52cb46118 100644
--- a/xen/arch/x86/spec_ctrl.c
+++ b/xen/arch/x86/spec_ctrl.c
@@ -16,11 +16,16 @@
  *
  * Copyright (c) 2017-2018 Citrix Systems Ltd.
  */
+#include <xen/cpu.h>
 #include <xen/errno.h>
 #include <xen/init.h>
 #include <xen/lib.h>
+#include <xen/stop_machine.h>
 #include <xen/warning.h>
 
+#include <public/platform.h>
+
+#include <asm/delay.h>
 #include <asm/microcode.h>
 #include <asm/msr.h>
 #include <asm/processor.h>
@@ -1266,6 +1271,174 @@ void __init init_speculation_mitigations(void)
         wrmsrl(MSR_MCU_OPT_CTRL, default_xen_mcu_opt_ctrl);
 }
 
+#define MICROCODE_CALLIN_TIMEOUT_US 30000
+#define MICROCODE_UPDATE_TIMEOUT_US 1000000
+
+static atomic_t cpu_in, cpu_out;
+static atomic_t next_cpu;
+
+static unsigned int common_caps;
+
+#define XENPF_spec_ctrl_success 0 /* New mitigations have been applied.       */
+                                  /* This is returned in case if some H/W     */
+                                  /* feature-bits have disappered! (Note that */
+                                  /* we don't expect this in practice.        */
+#define XENPF_spec_ctrl_noop    1 /* All mitigations are up to date */
+#define XENPF_spec_ctrl_error   2 /* Some error has occured */
+static uint32_t spec_ctrl_status;
+
+/* Wait for CPUs to rendezvous with a timeout (us) */
+static int wait_for_cpus(atomic_t *cnt, unsigned int expect,
+                         unsigned int timeout)
+{
+    while ( atomic_read(cnt) < expect )
+    {
+        if ( timeout <= 0 )
+        {
+            printk("CPU%d: Timeout when waiting for CPUs calling in\n",
+                   smp_processor_id());
+            return -EBUSY;
+        }
+        udelay(1);
+        timeout--;
+    }
+
+    return 0;
+}
+
+static int update_x86_caps(void)
+{
+    int cpu = smp_processor_id();
+    struct cpuinfo_x86 *c = &cpu_data[cpu];
+    unsigned int edx, tmp;
+
+    if ( c->cpuid_level < 0x00000007 )
+        return -ENOSYS;
+
+    /* Intel-defined CPU features, CPUID level 0x00000007:0.edx, word 9 */
+    cpuid_count(0x00000007, 0, &tmp, &tmp, &tmp, &edx);
+    c->x86_capability[cpufeat_word(X86_FEATURE_AVX512_4VNNIW)] = edx;
+
+    /* Find the new common feature subset */
+    if ( common_caps == 0 )
+        common_caps = edx;
+    else
+        common_caps &= edx;
+
+    return 0;
+}
+
+static int recalc_cpuid_policy(void)
+{
+    unsigned int idx = cpufeat_word(X86_FEATURE_AVX512_4VNNIW);
+
+    if ( boot_cpu_data.x86_capability[idx] == common_caps )
+    {
+        /* No new H/W features have been detected */
+        spec_ctrl_status = XENPF_spec_ctrl_noop;
+        return 0;
+    }
+
+    /* Update the global common feature set */
+    boot_cpu_data.x86_capability[idx] = common_caps;
+
+    /* Recalculate CPUID policies */
+    init_guest_cpuid();
+
+    spec_ctrl_status = XENPF_spec_ctrl_success;
+
+    return 0;
+}
+
+static int do_x86_caps_update(void *unused)
+{
+    int cpu = smp_processor_id();
+    unsigned int cpu_nr = num_online_cpus();
+    unsigned int finished;
+    int ret;
+    static bool error;
+
+    atomic_inc(&cpu_in);
+    ret = wait_for_cpus(&cpu_in, cpu_nr, MICROCODE_CALLIN_TIMEOUT_US);
+    if ( ret )
+        return ret;
+
+    while ( cpu != atomic_read(&next_cpu) )
+    {
+        finished = atomic_read(&cpu_out);
+        if ( wait_for_cpus(&next_cpu, cpu, MICROCODE_UPDATE_TIMEOUT_US) )
+        {
+            if ( atomic_read(&cpu_out) > finished )
+                continue;
+            printk("Timeout during do_x86_caps_update (finished %d/%d)",
+                   finished, cpu_nr);
+            return -EBUSY;
+        }
+    }
+
+    update_x86_caps();
+
+    finished = atomic_read(&cpu_out);
+    if ( finished == cpu_nr - 1 )
+    {
+        /* This is the last CPU. Update CPUID policy */
+        recalc_cpuid_policy();
+    }
+
+    atomic_set(&next_cpu, cpumask_next(cpu, &cpu_online_map));
+
+    atomic_inc(&cpu_out);
+    finished = atomic_read(&cpu_out);
+    while ( !error && finished != cpu_nr )
+    {
+        /*
+         * During each timeout interval, at least a CPU is expected to
+         * finish its update. Otherwise, something goes wrong.
+         */
+        if ( wait_for_cpus(&cpu_out, finished + 1,
+                           MICROCODE_UPDATE_TIMEOUT_US) && !error )
+        {
+            error = true;
+            printk("Timeout when finishing do_x86_caps_update (finished %d/%d)",
+                   finished, cpu_nr);
+            return -EBUSY;
+        }
+
+        finished = atomic_read(&cpu_out);
+    }
+
+    return 0;
+}
+
+long spec_ctrl_do_update(void *data)
+{
+    long ret = 0;
+
+    if ( !get_cpu_maps() )
+        return -EBUSY;
+
+    spec_ctrl_status = XENPF_spec_ctrl_error;
+
+    atomic_set(&cpu_in, 0);
+    atomic_set(&cpu_out, 0);
+    atomic_set(&next_cpu, cpumask_first(&cpu_online_map));
+
+    common_caps = 0;
+
+    ret = stop_machine_run(do_x86_caps_update, NULL, NR_CPUS);
+    if ( !ret )
+    {
+        if ( spec_ctrl_status == XENPF_spec_ctrl_error )
+            ret = -EFAULT;
+        if ( spec_ctrl_status == XENPF_spec_ctrl_noop )
+            ret = -ENOEXEC;
+    }
+
+    put_cpu_maps();
+
+    return ret;
+}
+
 static void __init __maybe_unused build_assertions(void)
 {
     /* The optimised assembly relies on this alias. */
diff --git a/xen/arch/x86/sysctl.c b/xen/arch/x86/sysctl.c
index af617464cdad..b44f821c64db 100644
--- a/xen/arch/x86/sysctl.c
+++ b/xen/arch/x86/sysctl.c
@@ -32,6 +32,7 @@
 #include <xsm/xsm.h>
 #include <asm/psr.h>
 #include <asm/cpuid.h>
+#include <asm/spec_ctrl.h>
 
 const struct cpu_policy system_policies[] = {
     [ XEN_SYSCTL_cpu_policy_raw ] = {
@@ -452,6 +453,19 @@ long arch_do_sysctl(
         break;
     }
 
+    case XEN_SYSCTL_spec_ctrl:
+    {
+        if ( sysctl->u.spec_ctrl.op != XENPF_spec_ctrl_update )
+        {
+            ret = -EINVAL;
+            break;
+        }
+
+        ret = continue_hypercall_on_cpu(smp_processor_id(),
+                                        spec_ctrl_do_update, NULL);
+        break;
+    }
+
     default:
         ret = -ENOSYS;
         break;
diff --git a/xen/include/asm-x86/spec_ctrl.h b/xen/include/asm-x86/spec_ctrl.h
index 59bab1a41bb6..461ee2f2ce7d 100644
--- a/xen/include/asm-x86/spec_ctrl.h
+++ b/xen/include/asm-x86/spec_ctrl.h
@@ -136,6 +136,8 @@ static always_inline void spec_ctrl_exit_idle(struct cpu_info *info)
      */
 }
 
+long spec_ctrl_do_update(void *data);
+
 #endif /* __ASSEMBLY__ */
 #endif /* !__X86_SPEC_CTRL_H__ */
 
diff --git a/xen/include/public/sysctl.h b/xen/include/public/sysctl.h
index d5d5976af5bc..18b44cc65805 100644
--- a/xen/include/public/sysctl.h
+++ b/xen/include/public/sysctl.h
@@ -1035,6 +1035,11 @@ typedef struct xen_sysctl_cpu_policy xen_sysctl_cpu_policy_t;
 DEFINE_XEN_GUEST_HANDLE(xen_sysctl_cpu_policy_t);
 #endif
 
+struct xen_sysctl_spec_ctrl {
+#define XENPF_spec_ctrl_update  0 /* Try to use new mitigations */
+    uint32_t op;      /* IN */
+};
+
 struct xen_sysctl {
     uint32_t cmd;
 #define XEN_SYSCTL_readconsole                    1
@@ -1065,6 +1070,7 @@ struct xen_sysctl {
 #define XEN_SYSCTL_livepatch_op                  27
 #define XEN_SYSCTL_set_parameter                 28
 #define XEN_SYSCTL_get_cpu_policy                29
+#define XEN_SYSCTL_spec_ctrl                    188
     uint32_t interface_version; /* XEN_SYSCTL_INTERFACE_VERSION */
     union {
         struct xen_sysctl_readconsole       readconsole;
@@ -1096,6 +1102,7 @@ struct xen_sysctl {
 #if defined(__i386__) || defined(__x86_64__)
         struct xen_sysctl_cpu_policy        cpu_policy;
 #endif
+        struct xen_sysctl_spec_ctrl         spec_ctrl;
         uint8_t                             pad[128];
     } u;
 };
